content/en/tracing/manual_instrumentation/ruby.md

```diff
@@ -1,23 +1,130 @@
       tag: 'Documentation'
       text: 'Connect your Logs and Traces together'
-    - link: 'tracing/opentracing'
-      tag: 'Documentation'
-      text: 'Implement Opentracing across your applications'
     - link: 'tracing/visualization/'
       tag: 'Documentation'
       text: 'Explore your services, resources, and traces'
 ---
+<div class="alert alert-info">
+If you have not yet read the instructions for auto-instrumentation and setup, please start with the <a href="https://docs.datadoghq.com/tracing/setup/ruby/">Ruby Setup Instructions</a>.
+</div>
+
+This page details common use cases for adding and customizing observability with Datadog APM.
+
+## Adding Tags
+
+Add custom [span tags][1] to your [spans][2] to customize your observability within Datadog.  The span tags are applied to your incoming traces, allowing you to correlate observed behavior with code-level information such as merchant tier, checkout amount, or user ID.
 
-If you aren't using supported library instrumentation (see [library compatibility][1]), you may want to to manually instrument your code. Adding tracing to your code is easy using the `Datadog.tracer.trace` method, which you can wrap around any Ruby code:
+### Add custom span tags
+
+Add custom tags to your spans corresponding to any dynamic value within your application code such as `customer.id`.
+
+{{< tabs >}}
+{{% tab "Active Span" %}}
+Access the current active [span][1] from any method within your code. **Note**: If the method is called and there is no span currently active, `active_span` is `nil`.
+
+```ruby
+require 'ddtrace'
+
+# get '/shopping_cart/:customer_id', to: 'shopping_cart#index'
+class ShoppingCartController < ApplicationController
+  # GET /shopping_cart
+  def index
+    # Get the active span
+    current_span = Datadog.tracer.active_span
+    # customer_id -> 254889
+    current_span.set_tag('customer.id', params.permit([:customer_id])) unless current_span.nil?
+
+    # [...]
+  end
+
+  # POST /shopping_cart
+  def create
+    # [...]
+  end
+end
+```
+
+[1]: /tracing/visualization/#spans
+{{% /tab %}}
+
+{{% tab "Manually Instrumented Spans" %}}
+
+Add [tags][1] directly to `Datadog::Span` objects by calling `#set_tag`:
+
+```ruby
+# An example of a Sinatra endpoint,
+# with Datadog tracing around the request.
+get '/posts' do
+  Datadog.tracer.trace('web.request') do |span|
+    span.set_tag('http.url', request.path)
+    span.set_tag('<TAG_KEY>', '<TAG_VALUE>')
+  end
+end
+```
+
+
+[1]: /tracing/visualization/#span-tags
+{{% /tab %}}
+{{< /tabs >}}
+
+### Adding tags globally to all spans
+
+Add [tags][1] to all [spans][2] by configuring the tracer with the `tags` option:
+
+```ruby
+Datadog.configure do |c|
+  c.tracer tags: { 'env' => 'dev' }
+end
+```
+
+You can also use the `DD_TAGS` environment variable to set tags on all spans for an application. For more information on Ruby environment variables, refer to the [setup documentation][3].
+
+### Set errors on a span
+
+TODO
```

Comment:
TODO? Did you have a code snippet for this or did you need one from me?

Reply:
Need from you

Reply:
@andrewardito @delner something like this ok?

-----

### Setting Errors on a Span

There are two ways to set an error on a span

- The first is to simply call `span.set_error` and pass in the Exception Object.
- This will automatically extract the error type, message, and backtrace.

```ruby
require 'ddtrace'
require 'timeout'

def example_method
  span = Datadog.tracer.trace('example.trace')
  puts 'some work'
  sleep(1)
  raise StandardError.new "This is an exception"
rescue StandardError => error
  span = Datadog.tracer.active_span
  span.set_error(error) unless span.nil?
  raise error
ensure
  span.finish
end

example_method()
```

- The second is to use `tracer.trace` which will by default set the error type, message, and backtrace. 
- To configure this behavior you can use the `on_error` option, which is the Handler invoked when a block is provided to `trace`, and the block raises an error. 
- The Proc is provided `span` and `error` as arguments. 
- By default, `on_error` Sets error on the span.

Default Behavior: `on_error`

```ruby
require 'ddtrace'
require 'timeout'

def example_method
  puts 'some work'
  sleep(1)
  raise StandardError.new "This is a exception"
end

Datadog.tracer.trace('example.trace', on_error: custom_error_handler) do |span|
  example_method()
end
```

Custom Behavior: `on_error`

```ruby
require 'ddtrace'
require 'timeout'

def example_method
  puts 'some work'
  sleep(1)
  raise StandardError.new "This is a special exception"
end

custom_error_handler = proc do |span, error|
  span.set_tag('custom_tag', 'custom_value')
  span.set_error(error) unless error.message.include?("a special exception")
end

Datadog.tracer.trace('example.trace', on_error: custom_error_handler) do |span|
  example_method()
end  
```

Reply:
🙇 Thank you! @delner Can you give a 👍  and then we're good to merge.